sdk/servicebus/service-bus/src/serviceBusAtomManagementClient.ts

```diff
@@ -0,0 +1,942 @@
+
+/**
+ * Update Rule response
+ */
+export type UpdateRuleResponse = RuleResponse;
+
+/**
+ * Delete Rule response
+ */
+export type DeleteRuleResponse = RuleResponse;
+
+/**
+ * Represents result of list operation on rules.
+ */
+export interface ListRulesResponse extends Array<Rule> {
+  /**
+   * The underlying HTTP response.
+   */
+  _response: HttpOperationResponse;
+}
+
+/**
+ * All operations return promises that resolve to an object that has the relevant output.
+ * These objects also have a property called `_response` that you can use if you want to
+ * access the direct response from the service.
+ */
+export class ServiceBusAtomManagementClient extends ServiceClient {
+  private endpoint: string;
+
+  private queueResourceSerializer: AtomXmlSerializer;
+  private topicResourceSerializer: AtomXmlSerializer;
+  private subscriptionResourceSerializer: AtomXmlSerializer;
+  private ruleResourceSerializer: AtomXmlSerializer;
+
+  /**
+   * Initializes a new instance of the ServiceBusManagementClient class.
+   * @param connectionString The connection string needed for the client to connect to Azure.
+   * @param options ServiceBusAtomManagementClientOptions
+   */
+  constructor(connectionString: string, options?: ServiceBusAtomManagementClientOptions) {
+    const connectionStringObj = ServiceBusAtomManagementClient.parseConnectionString(
+      connectionString
+    );
+
+    if (connectionStringObj.Endpoint == undefined) {
+      throw new Error("Endpoint must be supplied in the connection string");
+    }
+
+    const credentials = new SasServiceClientCredentials(
+      connectionStringObj.SharedAccessKeyName,
+      connectionStringObj.SharedAccessKey
+    );
+
+    const requestPolicyFactories: RequestPolicyFactory[] = [];
+    requestPolicyFactories.push(logPolicy(log.httpAtomXml));
+    requestPolicyFactories.push(signingPolicy(credentials));
+
+    if (options && options.proxySettings) {
+      requestPolicyFactories.push(proxyPolicy(options.proxySettings));
+    }
+    const serviceClientOptions: ServiceClientOptions = {
+      requestPolicyFactories: requestPolicyFactories
+    };
+
+    super(credentials, serviceClientOptions);
+    this.queueResourceSerializer = new QueueResourceSerializer();
+    this.topicResourceSerializer = new TopicResourceSerializer();
+    this.subscriptionResourceSerializer = new SubscriptionResourceSerializer();
+    this.ruleResourceSerializer = new RuleResourceSerializer();
+    this.endpoint = (connectionString.match("Endpoint=sb://(.*)/;") || "")[1];
+  }
+
+  /**
+   * Creates a queue with given name, configured using the given options
+   * @param queueName
+   * @param queueOptions Options to configure the Queue being created.
+   * For example, you can configure a queue to support partitions or sessions
+   */
+  async createQueue(queueName: string, queueOptions?: QueueOptions): Promise<CreateQueueResponse> {
+    if (!queueName.match(validQueueOrTopicNameRegex)) {
```

Comment:
https://github.com/Azure/azure-sdk-for-js/pull/4922#discussion_r337155826
Added in comments as well to help with reasoning

Reply:
Did you mean comments in code? I checked the latest code, I couldn't find any comments around `createQueue()`, `createTopic()` or where the regex is declared.


> Service allows / in queue and topic names, but not in namespace, subscription and rules 

The above quote is copy pasted from https://github.com/Azure/azure-sdk-for-js/pull/4922#discussion_r337155826 which was linked above for reference:

This still doesn't explain why we need to validate the name for queues & topics. There is no precedence for client sdks validating entity names as the service can change its behavior any day. Therefore, we defer the responsibility of name validation to the service.

cc @bterlson if he has more to add to this.

Reply:
Agreed, but since the URL parsing and construction is something that the client is doing (current feature in this case), we should factor in the service driven constraints I believe, because otherwise bugs would get introduced.
For instance, topic name can be "abcd/Subscriptions/" and this would cause URL parsing logic to fail.

About the regex, that was picked up from Azure portal prompt.

Reply:
Oops, the comment was for why we were "depending" on the name validation.
About the check in itself, it gives a better user experience and more meaningful error message instead of a failed request that just says 400 Bad Request. 
Also, we would anyway have a hard dependency on the service driven naming constraints.
So I'd keep this check - so that the feature can be updated should anything change on service side.

Reply:
Invalid names for subscriptions & rules dont get similar 400 Bad Request from the service?
Does the service give an error in these cases or just allow such names to be valid?

Reply:
It does give a 400 error with some error details - the message is not entirely accurate (about use of `/` in particular) but seems good enough. Removed the name checks

Reply:
Sample error message returned by service - " ... contains character(s) that is not allowed by Service Bus. Entity segments can contain only letters, numbers, periods (.), hyphens (-), and underscores (_). ... "

Reply:
I would keep our error message similar to what the service returns for subscriptions & rules.
We follow up in the issue logged to review the error messages